server/types/interval.go

```diff
@@ -0,0 +1,265 @@
+	}
+	bc, _, err := b.Convert(v2)
+	if err != nil {
+		return 0, err
+	}
+
+	ab := ac.(duration.Duration)
+	bb := bc.(duration.Duration)
+	return ab.Compare(bb), nil
+}
+
+// Convert implements the DoltgresType interface.
+func (b IntervalType) Convert(val any) (any, sql.ConvertInRange, error) {
+	switch val := val.(type) {
+	case duration.Duration:
+		return val, sql.InRange, nil
+	case nil:
+		return nil, sql.InRange, nil
+	default:
+		return nil, sql.OutOfRange, fmt.Errorf("%s: unhandled type: %T", b.String(), val)
+	}
+}
+
+// Equals implements the DoltgresType interface.
+func (b IntervalType) Equals(otherType sql.Type) bool {
+	if otherExtendedType, ok := otherType.(types.ExtendedType); ok {
+		return bytes.Equal(MustSerializeType(b), MustSerializeType(otherExtendedType))
+	}
+	return false
+}
+
+// FormatValue implements the DoltgresType interface.
+func (b IntervalType) FormatValue(val any) (string, error) {
+	if val == nil {
+		return "", nil
+	}
+	return b.IoOutput(sql.NewEmptyContext(), val)
+}
+
+// GetSerializationID implements the DoltgresType interface.
+func (b IntervalType) GetSerializationID() SerializationID {
+	return SerializationID_Interval
+}
+
+// IoInput implements the DoltgresType interface.
+func (b IntervalType) IoInput(ctx *sql.Context, input string) (any, error) {
+	dInterval, err := tree.ParseDInterval(input)
+	if err != nil {
+		return nil, err
+	}
+	return dInterval.Duration, nil
+}
+
+// IoOutput implements the DoltgresType interface.
+func (b IntervalType) IoOutput(ctx *sql.Context, output any) (string, error) {
+	converted, _, err := b.Convert(output)
+	if err != nil {
+		return "", err
+	}
+	// TODO: depends on `intervalStyle` configuration variable. Defaults to `postgres`.
+	d := converted.(duration.Duration)
+	return d.String(), nil
+}
+
+// IsPreferredType implements the DoltgresType interface.
+func (b IntervalType) IsPreferredType() bool {
+	return true
+}
+
+// IsUnbounded implements the DoltgresType interface.
+func (b IntervalType) IsUnbounded() bool {
+	return false
+}
+
+// MaxSerializedWidth implements the DoltgresType interface.
+func (b IntervalType) MaxSerializedWidth() types.ExtendedTypeSerializedWidth {
+	return types.ExtendedTypeSerializedWidth_64K
+}
+
+// MaxTextResponseByteLength implements the DoltgresType interface.
+func (b IntervalType) MaxTextResponseByteLength(ctx *sql.Context) uint32 {
+	return 64
+}
+
+// OID implements the DoltgresType interface.
+func (b IntervalType) OID() uint32 {
+	return uint32(oid.T_interval)
+}
+
+// Promote implements the DoltgresType interface.
+func (b IntervalType) Promote() sql.Type {
+	return Interval
+}
+
+// SerializedCompare implements the DoltgresType interface.
+func (b IntervalType) SerializedCompare(v1 []byte, v2 []byte) (int, error) {
+	if len(v1) == 0 && len(v2) == 0 {
+		return 0, nil
+	} else if len(v1) > 0 && len(v2) == 0 {
+		return 1, nil
+	} else if len(v1) == 0 && len(v2) > 0 {
+		return -1, nil
+	}
+
+	return bytes.Compare(v1, v2), nil
```

Comment:
This comparison does not mimic the same comparison in `Compare`. This one works on the serialized form, which is a string, while `Compare` uses the actual `Duration` types. We should mimic the one that Postgres is doing.

Reply:
Seems like we can and should do better than a string for the serialized form. Would be great if the binary format of the serialized form were directly comparable / sortable without parsing.

Reply:
Would using JSON for serializing this struct be better?

Reply:
It wouldn't be any better, JSON is also a string so it'd be the same. Looking at `duration.Duration`, it has `Encode` and `Decode` functions, so we can use `utils.Writer` to write them in `nano, month, day` order. That would allow us to use `bytes.Equal` in `SerializedCompare`.

Reply:
I wouldn't use JSON either. You want to have a format that can sort natively by comparing the bytes without any parsing. This is how postgres serializes this type:

```
Internally, interval values are stored as three integral fields: months, days, and microseconds. These fields are kept separate because the number of days in a month varies, while a day can have 23 or 25 hours if a daylight savings time transition is involved. An interval input string that uses other units is normalized into this format, and then reconstructed in a standardized way for output, for example:

SELECT '2 years 15 months 100 weeks 99 hours 123456789 milliseconds'::interval;
               interval
---------------------------------------
 3 years 3 mons 700 days 133:17:36.789
```

So you want basically 3 things in this order:

* an int64 for total microseconds of this duration, for sort order (assume 24 hours in a day and 30 days in a month for this, if those fields are specified)
* an int32 for number of months
* an int32 for number of days

This gives us a fixed width storage size of 8 bytes, same as postgres.

When decoding, you get the microseconds by substracting the number of months + days from the sort field again.

Reply:
Zach has essentially said the same thing in more words lol